lib/plugins/cpuUsageThrottle.js

```diff
@@ -0,0 +1,139 @@
+ *    your process to put more emphasis on recent historical CPU usage when
+ *    determininng whether it should shed load, set this to a higher value. The
+ *    unit is in ms. The default is to set this to the same value as interval.
+ * @param {Error} [opts.err] A restify error used as a response when the
+ *    cpu usage limit is exceeded
+ * @returns {Function} middleware to be registered on server.pre
+ */
+function cpuUsageThrottle (opts) {
+
+    // Scrub input and populate our configuration
+    assert.object(opts, 'opts');
+    assert.number(opts.limit, 'opts.limit');
+    assert.optionalNumber(opts.max, 'opts.max');
+    assert.number(opts.interval, 'opts.interval');
+    assert.optionalNumber(opts.halfLife, 'opts.halfLife');
+
+    if (opts.err !== undefined && opts.err !== null) {
+        assert.ok(opts.err instanceof Error, 'opts.err must be an error');
+        assert.optionalNumber(opts.err.statusCode, 'opts.err.statusCode');
+    }
+
+    var self = {};
+    self._err = opts.err || defaultResponse;
+    self._limit = opts.limit;
+    self._max = opts.max || 1;
+    self._interval = opts.interval;
+    self._halfLife = (typeof opts.halfLife === 'number') ?
+      opts.halfLife : opts.interval;
+    self._timeout = null;
+    assert.ok(self._max > self._limit, 'limit must be less than max');
+
+    // We start at 0 to compensate for the initial spike in CPU usage a server
+    // experiences when starting up.
+    self._ewma = new EWMA(self._halfLife, 0);
+
+    // self._reject represents the % of traffic that we should reject at the
+    // current point in time based on how much over our limit we are. This is
+    // updated on an interval by updateReject().
+    self._reject = 0;
+
+    // updateReject should be called on an interval, it checks the average CPU
+    // usage between two invocations of updateReject.
+    function updateReject() {
+        pidusage.stat(process.pid, function (e, stat) {
+            // If we fail to stat for whatever reason, err on the side of
+            // accepting traffic, while failure probably indicates something
+            // horribly wrong with the server, identifying this error case and
+            // appropriately responding is way outside the scope of this plugin
+            if (e) {
+                // When cpu is NaN, we should never reject
+                stat = { cpu: NaN };
+            }
+
+            // Divide by 100 to match linux's top format
+            self._cpu = stat.cpu / 100;
+            // Update reject with the % of traffic we should be rejecting. This
+            // is safe since max > limit so it can never be division by 0. If
+            // the current cpu usage is less that the limit, _reject will be
+            // negative and we will never shed load
+            self._reject =
+                (self._cpu - self._limit) / (self._max - self._limit);
+            self._timeout = setTimeout(updateReject, self._interval);
```

Comment:
Under heavy load `updateReject()` might execute later than you'd like it to - is it valuable to incorporate that delta into the algo?

Reply:
Yup, EWMA accounts for this :-)

Reply:
It is the reason for the crazy magic maths here: https://github.com/ReactiveSocket/ewma/blob/master/index.js#L55

Reply:
Even tho ewma accounts for this in the algorithm, the interval is a parameter configured by the user, and we should strive to stick to said interval. I think it still makes sense to keep track of the delta here, since changes to the interval will affect the throttling algorithm.

Reply:
> Even tho ewma accounts for this in the algorithm, the interval is a parameter configured by the user, and we should strive to stick to said interval.

Agreed. That is what I'm trying to accomplish with this code, though there may be a better way. Since `pid.stat` is async and it's performance is not necessarily bound to process saturation, its possible the first call to it could take a few minutes to return. With `setInterval` we have the potential to queue a whole bunch of `pid.stat` calls up while that first blocks. Then it ends up being a race to burn through that queue. This tries to ensure we call `pid.stat` exactly once at a time and that the time between it returning back to us and us calling it again is as close to the interval the user requested as possible.

Reply:
> keep track of the delta

That would be valuable, though I'm not sure how we could reliably expose that metric.

Some ideas:

* We could attach it to `self`, but whatever logs `self` at the end of the day won't be bound to this interval.
* We could emit an event, but that would increase the surface area of our plugin API and would probably be something we want to generalize. (also worth noting that `plugin.state` in this PR is doing this already)

Reply:
> changes to the interval will affect the throttling algorithm

While it may change the calculation slightly, if it has a meaningful impact on the behaviour of throttling we will need revisit this implementation. It is currently designed to be stable in the face of unevenly spaced data points, and since our data points actually represent several million values (either 1 or 0 indicating whether Node used a cycle of the CPU) averaged over the interval, every "point" is still meaningfully represented in the calculation irrespective of latency. The decay of all of those "points" represented in a single average will be the same as if they all occurred at the same moment in time. That being said, from an external perspective, this _should_ result in nearly identical behaviour. Like I said, if this proves to not be the case, we will need to re-think this implementation.

Worth noting that in order to measure whether this has any "meaningful impact" we need to log the delta somewhere :-P

Reply:
It is probably worth clarifying why `interval` is configurable if there isn't a significant impact of a high delta between when the user requested interval and the actual time between checks due to resource contention.

While I don't believe contention increasing this delta will have a meaningful impact on the behaviour of load shedding, I do believe that, during periods where there isn't contention for resources, a more frequent check may be desirable and will have a meaningful impact. So essentially, the user may want to _lower_ this value to be able to measure CPU spikes with a higher resolution.

Reply:
What about just exposing the delta for now then (through EE or similar) and we can keep track of it as we first deploy? As we gather data it should inform us just how much of an impact there will be.

Reply:
In other words, no need to consume it in a meaningful way now, but at least we have some visibility into what's going on. Although, now that I think about it - would something like this already be captured in the event loop metrics?

Reply:
Sorry, I think there's been some confusion. I am suggesting we keep track of how much time has elapsed so that we can update `interval` appropriately. e.g. if the interval is 500ms, but it took us 300ms to go through this current interval, we should then set timeout to 200ms, so that we're ensuring we're firing the interval every 500ms. Otherwise in this example we would be running the next interval 800ms since the last interval.

Reply:
There could be latency introduced by the file descriptor read on Linux or spinning up the subprocess on windows (inside or `pidusage.stat`). I'll export it as a value off of `state`

Reply:
Also adding it to the error context when we shed load